timer/src/main/java/com/toggl/timer/startedit/domain/UpdateAutocompleteSuggestionsEffect.kt

```diff
@@ -53,13 +63,42 @@ class UpdateAutocompleteSuggestionsEffect(
             return task.name.contains(word, true)
         }
 
+        val words = query.split(wordSeparator)
+
         return words.fold(timeEntries.values) { timeEntries, word ->
             timeEntries.filter { timeEntry ->
                 timeEntry.description.contains(word, true) ||
-                timeEntry.projectOrClientNameContains(word) ||
-                timeEntry.tagNamesContain(word) ||
-                timeEntry.taskNameContains(word)
+                    timeEntry.projectOrClientNameContains(word) ||
+                    timeEntry.tagNamesContain(word) ||
+                    timeEntry.taskNameContains(word)
             }
         }.map(AutocompleteSuggestion::TimeEntry)
     }
+
+    private fun fetchProjectSuggestionsFor(query: String): List<AutocompleteSuggestion> {
+        fun Project.clientNameContains(word: String): Boolean {
+            val client = clientId?.run(clients::get) ?: return false
+            return client.name.contains(word, true)
+        }
+
+        val words = query.split(wordSeparator)
+        return sequence {
+            yield(AutocompleteSuggestion.CreateProject(query))
+            words.fold(projects.values) { projects, word ->
+                projects.filter { project ->
+                    project.name.contains(word, true) ||
+                        project.clientNameContains(word)
+                }
+            }.map(AutocompleteSuggestion::Project).let { yieldAll(it) }
+        }.toList()
+    }
+
+    private fun fetchTasksSuggestionsFor(query: String): List<AutocompleteSuggestion> {
+        val words = query.split(wordSeparator)
+        return words.fold(tasks.values) { tasks, word ->
+            tasks.filter { task ->
+                task.name.contains(word, true)
+            }
+        }.map(AutocompleteSuggestion::Task)
+    }
```

Comment:
These fetch methods are starting to look very similar. What about defining helper method to hide some inner plumbing in order to simplify our fetch methods and make them more readable. 

```suggestion
    private fun fetchTasksSuggestionsFor(query: String): List<AutocompleteSuggestion> =
        fetchSuggestionsFor(query, tasks.values, AutocompleteSuggestion::Task) { task, word ->
            task.name.contains(word, true)
        }
        
    private fun <T> fetchSuggestionsFor(
        query: String,
        possibleSuggestions: Collection<T>,
        toSuggestion: (T) -> AutocompleteSuggestion,
        predicate: (T, String) -> Boolean
    ): List<AutocompleteSuggestion> {
        val words = query.split(wordSeparator)
        return words.fold(possibleSuggestions) { remainingPossibleSuggestions, word ->
            remainingPossibleSuggestions.filter { predicate(it, word) }
        }.map(toSuggestion)
    }
```